index.js

```diff
@@ -1,73 +1,111 @@
 'use strict';
+/* global Map:readonly, Set:readonly, ArrayBuffer:readonly */
 
-var isArray = Array.isArray;
-var keyList = Object.keys;
-var hasProp = Object.prototype.hasOwnProperty;
 var hasElementType = typeof Element !== 'undefined';
+var hasMap = typeof Map === 'function';
+var hasSet = typeof Set === 'function';
+var hasArrayBuffer = typeof ArrayBuffer === 'function';
+
+// Note: We **don't** need `envHasBigInt64Array` in fde es6/index.js
 
 function equal(a, b) {
-  // fast-deep-equal index.js 2.0.1
+  // START: fast-deep-equal es6/index.js 3.1.1
   if (a === b) return true;
 
   if (a && b && typeof a == 'object' && typeof b == 'object') {
-    var arrA = isArray(a)
-      , arrB = isArray(b)
-      , i
-      , length
-      , key;
+    if (a.constructor !== b.constructor) return false;
 
-    if (arrA && arrB) {
+    var length, i, keys;
+    if (Array.isArray(a)) {
       length = a.length;
       if (length != b.length) return false;
       for (i = length; i-- !== 0;)
         if (!equal(a[i], b[i])) return false;
       return true;
     }
 
-    if (arrA != arrB) return false;
+    // START: Modifications:
+    // 1. Extra `has<Type> &&` helpers in initial condition allow es6 code
+    //    to co-exist with es5.
+    // 2. Replace `for of` with es5 compliant iteration using `for`.
+    //    Basically, take:
+    //
+    //    ```js
+    //    for (i of a.entries())
+    //      if (!b.has(i[0])) return false;
+    //    ```
+    //
+    //    ... and convert to:
+    //
+    //    ```js
+    //    it = a.entries();
+    //    for (i = it.next(); !i.done; i = it.next())
+    //      if (!b.has(i.value[0])) return false;
+    //    ```
+    //
+    //    **Note**: `i` access switches to `i.value`.
+    var it;
+    if (hasMap && (a instanceof Map) && (b instanceof Map)) {
+      if (a.size !== b.size) return false;
+      it = a.entries();
+      for (i = it.next(); !i.done; i = it.next())
+        if (!b.has(i.value[0])) return false;
+      it = a.entries();
+      for (i = it.next(); !i.done; i = it.next())
+        if (!equal(i.value[1], b.get(i.value[0]))) return false;
```

Comment:
We could do both checks above to avoid the additional iteration

Reply:
They're split upstream.

I'm guessing the reason upstream splits is the first iteration is non-recursive, but this one _is_ recursive, which can be a lot more expensive.

Reply:
Oh true! Didn’t think of the function call cost here